lib/services/systemd.c

```diff
@@ -828,43 +828,71 @@ systemd_timeout_callback(gpointer p)
     return FALSE;
 }
 
-/* For an asynchronous 'op', returns FALSE if 'op' should be free'd by the caller */
-/* For a synchronous 'op', returns FALSE if 'op' fails */
-gboolean
-systemd_unit_exec(svc_action_t * op)
+/*!
+ * \internal
+ * \brief Execute a systemd action
+ *
+ * \param[in] op  Action to execute
+ *
+ * \return Standard Pacemaker return code
+ * \retval EBUSY          Recurring operation could not be initiated
+ * \retval pcmk_rc_error  Synchronous action failed
+ * \retval pcmk_rc_ok     Synchronous action succeeded, or asynchronous action
+ *                        should not be freed (because it already was or is
+ *                        pending)
+ *
+ * \note If the return value for an asynchronous action is not pcmk_rc_ok, the
+ *       caller is responsible for freeing the action.
+ */
+int
+services__execute_systemd(svc_action_t *op)
 {
-    char *unit = NULL;
+    CRM_ASSERT(op != NULL);
+
+    if ((op->action == NULL) || (op->agent == NULL)) {
+        op->rc = PCMK_OCF_NOT_CONFIGURED;
+        op->status = PCMK_EXEC_ERROR_FATAL;
+        goto done;
+    }
+
+    if (!systemd_init()) {
+        op->rc = PCMK_OCF_UNKNOWN_ERROR;
+        op->status = PCMK_EXEC_ERROR;
+        goto done;
+    }
 
-    CRM_ASSERT(op);
-    CRM_ASSERT(systemd_init());
-    op->rc = PCMK_OCF_UNKNOWN_ERROR;
     crm_debug("Performing %ssynchronous %s op on systemd unit %s named '%s'",
-              op->synchronous ? "" : "a", op->action, op->agent, op->rsc);
+              (op->synchronous? "" : "a"), op->action, op->agent,
+              crm_str(op->rsc));
 
     if (pcmk__str_eq(op->action, "meta-data", pcmk__str_casei)) {
-        // @TODO Implement an async meta-data call in executor API
         op->stdout_data = systemd_unit_metadata(op->agent, op->timeout);
         op->rc = PCMK_OCF_OK;
-
-        if (op->synchronous == FALSE) {
-            return services__finalize_async_op(op) == pcmk_rc_ok;
-        }
-        return TRUE;
+        op->status = PCMK_EXEC_DONE;
+        goto done;
     }
 
-    unit = systemd_unit_by_name(op->agent, op);
-    free(unit);
+    // Initialize rc/status in case systemd_unit_by_name() doesn't set them
+    op->rc = PCMK_OCF_UNKNOWN_ERROR;
+    op->status = PCMK_EXEC_DONE;
 
-    if (op->synchronous == FALSE) {
-        if (op->opaque->pending) {
-            op->opaque->timerid = g_timeout_add(op->timeout + 5000, systemd_timeout_callback, op);
-            services_add_inflight_op(op);
-            return TRUE;
+    {
+        char *unit = systemd_unit_by_name(op->agent, op);
```

Comment:
good candidate for re-factoring/naming:
could be more obvious that is actually executes something and doesn't just get the unit.
2 out of 3 usages we right away free unit that was just strdupped.

Reply:
Definitely, that's in the next PR ;) wrapping up testing now